File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/TableSchema.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.flink.table.api

import org.apache.flink.table.calcite.FlinkTypeFactory
import org.apache.flink.table.api.types.{DataTypes, _}

import _root_.scala.collection.mutable.ArrayBuffer

/**
 * A TableSchema represents a Table's structure.
 */
class TableSchema(
    private val columns: Array[Column],
    private val primaryKey: Array[String],
    private val uniqueKeys: Array[Array[String]],
    private val computedColumns: Array[ComputedColumn],
    private val watermarks: Array[Watermark]) {

  val columnNameToIndex: Map[String, Int] = columns.zipWithIndex.map {
    case (column, index) => (column.name, index)
  }.toMap

  primaryKey.foreach {
    case field: String if columnNameToIndex.contains(field) =>
    case field: String => new TableException(s"Primary key field: $field not found in table schema")
  }

  uniqueKeys.foreach {
    case uniqueKey: Array[String] if uniqueKey.isEmpty =>
      new TableException("Unique key should not be empty.")
    case uniqueKey: Array[String] => uniqueKey.foreach {
      case field: String if columnNameToIndex.contains(field) =>
      case field: String =>
        new TableException(s"Unique key field: $field not found in table schema")
    }
  }

  def this(columns: Array[Column]) = {
    this(
      columns,
      new Array[String](0),
      new Array[Array[String]](0),
      new Array[ComputedColumn](0),
      new Array[Watermark](0))
  }

  def this(names: Array[String], types: Array[InternalType], nullables: Array[Boolean]) = {
    this {
      if (names.length != types.length) {
        throw new TableException(
          s"Number of column indexes and column names must be equal." +
              s"\nColumn names count is [${names.length}]" +
              s"\nColumn types count is [${types.length}]" +
              s"\nColumn names:${names.mkString("[ ", ", ", " ]")}" +
              s"\nColumn types:${types.mkString("[ ", ", ", " ]")}")
      }

      if (names.length != nullables.length) {
        throw new TableException(
          s"Number of column indexes and column names must be equal." +
              s"\nColumn names count is [${names.length}]" +
              s"\nColumn types count is [${nullables.length}]"
        )
      }

      // check uniqueness of field names
      if (names.toSet.size != types.length) {
        val columnNameBuffer = names.toBuffer
        val duplicate = names.filter(
          name => columnNameBuffer.-=(name).contains(name))

        throw new TableException(
          s"Table column names must be unique." +
              s"\nThe duplicate columns are: ${duplicate.mkString("[ ", ", ", " ]")}" +
              s"\nAll column names: ${names.mkString("[ ", ", ", " ]")}")
      }

      names.zip(types).zipWithIndex.map {
        case ((name, typeInfo), index) =>
          Column(name, typeInfo, nullables(index))
      }
    }
  }

  def this(names: Array[String], types: Array[InternalType]) = {
    this(names, types, types.map(
      {
        case typ: ExternalType => typ.toInternalType
        case typ: InternalType => typ
      }).map(
      typ =>
        typ!= TimestampType.ROWTIME_INDICATOR &&
            typ != TimestampType.PROCTIME_INDICATOR))
  }

  def getColumns: Array[Column] = columns

  def getPrimaryKeys: Array[String] = primaryKey

  def getUniqueKeys: Array[Array[String]] = uniqueKeys

  def getNullables: Array[Boolean] = columns.map(_.isNullable)

  def getTypes: Array[InternalType] = columns.map(_.internalType)

  def getColumnNames: Array[String] = columns.map(_.name)

  def getComputedColumns: Array[ComputedColumn] = computedColumns

  def getWatermarks: Array[Watermark] = watermarks

  /**
   * Returns the specified column for the given column index.
   *
   * @param columnIndex the index of the field
   */
  def getColumn(columnIndex: Int): Column = {
    require(columnIndex >= 0 && columnIndex < columns.length)
    columns(columnIndex)
  }

  /**
   * Returns the specified type information for the given column index.
   *
   * @param columnIndex the index of the field
   */
  def getType(columnIndex: Int): InternalType = {
    require(columnIndex >= 0 && columnIndex < columns.length)
    columns(columnIndex).internalType
  }

  /**
   * Returns the specified type information for the given column name.
   *
   * @param columnName the name of the field
   */
  def getType(columnName: String): Option[InternalType] = {
    if (columnNameToIndex.contains(columnName)) {
      Some(columns(columnNameToIndex(columnName)).internalType)
    } else {
      None
    }
  }

  /**
   * Returns the specified column name for the given column index.
   *
   * @param columnIndex the index of the field
   */
  def getColumnName(columnIndex: Int): String = {
    require(columnIndex >= 0 && columnIndex < columns.length)
    columns(columnIndex).name
  }

  /**
    * Converts a table schema into a schema that represents the result that would be written
    * into a table sink or operator outside of the Table & SQL API. Time attributes are replaced
    * by proper TIMESTAMP data types.
    *
    * @return a table schema with no time attributes
    */
  def withoutTimeAttributes: TableSchema = {
    val converted = columns.map { t =>
      if (FlinkTypeFactory.isTimeIndicatorType(t.internalType)) {
        Column(t.name, TimestampType.TIMESTAMP, false)
      } else {
        t
      }
    }
    new TableSchema(converted)
  }

  /**
    * Converts a table schema into a (nested) type information describing a Row.
    *
    * @return type information where columns are fields of a row
    */
  def toRowType: DataType = {
    DataTypes.createRowType(getTypes, getColumnNames)
  }

  override def toString: String = {
    val builder = new StringBuilder
    builder.append("root\n")
    columns.foreach {
      column: Column =>
        builder.append(s" |--    name: ${column.name}\n")
        builder.append(s"    |-- type: ${column.internalType}\n")
        builder.append(s"    |-- isNullable: ${column.isNullable}\n")
    }
    if (!primaryKey.isEmpty) {
      builder.append(s"primary key\n")
      builder.append(s" |--  ${primaryKey.mkString(",")}\n")
    }
    if (!uniqueKeys.isEmpty) {
      builder.append(s"unique keys\n")
          uniqueKeys.foreach {
            case uniqueKey: Array[String] =>
              builder.append(s" |--  ${uniqueKey.mkString(",")}\n")
          }
    }
    builder.toString()
  }

  override def equals(obj: Any): Boolean = {
    obj match {
      case t:TableSchema => toString.equals(t.toString)
      case _ => false
    }
  }

}

/**
 * Table Columns.
 */
case class Column(
    name: String, // column name
    internalType: InternalType, // physical type of column
    isNullable: Boolean = true  // is nullable
)

case class ComputedColumn(name: String, expression: String)

case class Watermark(name: String, eventTime: String, offset: Long)

object TableSchema {

  /**
   * Create table schema from a data type.
   */
  def fromDataType(
      dataType: DataType,
      fieldNullables: Option[Array[Boolean]] = None): TableSchema = {
    DataTypes.internal(dataType) match {
      case bt: BaseRowType =>
        val fieldNames = bt.getFieldNames
        val fieldTypes = bt.getFieldTypes
        if (fieldNullables.isDefined) {
          new TableSchema(fieldNames, fieldTypes, fieldNullables.get)
        } else {
          new TableSchema(fieldNames, fieldTypes)
        }
      case t =>
        val fieldNames = Array("f0")
        val fieldTypes = Array(t)
        if (fieldNullables.isDefined) {
          new TableSchema(fieldNames, fieldTypes, fieldNullables.get)
        } else {
          new TableSchema(fieldNames, fieldTypes)
        }
    }
  }

  def builder(): TableSchemaBuilder = {
    new TableSchemaBuilder
  }
}

class TableSchemaBuilder {

  private val columns: ArrayBuffer[Column] = new ArrayBuffer[Column]()
  private val primaryKey: ArrayBuffer[String] = new ArrayBuffer[String]()
  private val uniqueKeys: ArrayBuffer[ArrayBuffer[String]] = new ArrayBuffer[ArrayBuffer[String]]()
  private val computedColumns: ArrayBuffer[ComputedColumn] = new ArrayBuffer[ComputedColumn]()
  private val watermarks: ArrayBuffer[Watermark] = new ArrayBuffer[Watermark]()

  def fromDataType(dataType: DataType): TableSchemaBuilder = {
    columns.append(TableSchema.fromDataType(dataType).getColumns:_*)
    this
  }

  @deprecated
  def field(name: String, tpe: DataType): TableSchemaBuilder = {
    columns.append(Column(name, DataTypes.internal(tpe)))
    this
  }

  @deprecated
  def field(name: String, tpe: DataType, nullable: Boolean): TableSchemaBuilder = {
    columns.append(Column(name, DataTypes.internal(tpe), nullable))
    this
  }

  def column(name: String, tpe: DataType): TableSchemaBuilder = {
    columns.append(Column(name, DataTypes.internal(tpe)))
    this
  }

  def column(name: String, tpe: DataType, nullable: Boolean): TableSchemaBuilder = {
    columns.append(Column(name, DataTypes.internal(tpe), nullable))
    this
  }

  def computedColumn(name: String, expression: String): TableSchemaBuilder = {
    computedColumns.append(ComputedColumn(name, expression))
    this
  }

  def watermark(name: String, from: String, offset: Long): TableSchemaBuilder = {
    watermarks.append(Watermark(name, from, offset))
    this
  }

  def primaryKey(field: String*): TableSchemaBuilder = {
    field.foreach(f => primaryKey.append(f))
    this
  }

  def uniqueKey(field: String*): TableSchemaBuilder = {
    val newUniqueKey = new ArrayBuffer[String]()
    field.foreach(f => newUniqueKey.append(f))
    uniqueKeys.append(newUniqueKey)
    this
  }

  def build(): TableSchema = {
    new TableSchema(
      columns.toArray,
      primaryKey.toArray,
      uniqueKeys.map(_.toArray).toArray,
      computedColumns.toArray,
      watermarks.toArray
    )
  }
}

File: flink-libraries/flink-table-common/src/main/java/org/apache/flink/table/api/TableSchema.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.flink.table.api;

import org.apache.flink.annotation.PublicEvolving;
import org.apache.flink.table.api.types.InternalType;
import org.apache.flink.table.api.types.TimestampType;
import org.apache.flink.util.Preconditions;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.Set;
import java.util.stream.Collectors;

/**
  * A table schema that represents a table's structure with field names and types.
  */
@PublicEvolving
public class TableSchema {

	private final Column[] columns;

	private final String[] primaryKeys;

	private final String[][] uniqueKeys;

	private final ComputedColumn[] computedColumns;

	private final Watermark[] watermarks;

	private final Map<String, Integer> columnNameToIndex;

	public TableSchema(Column[] columns) {
		this(columns, new String[]{}, new String[][]{}, new ComputedColumn[]{}, new Watermark[]{});
	}

	public TableSchema(
			Column[] columns, String[] primaryKeys, String[][] uniqueKeys,
			ComputedColumn[] computedColumns, Watermark[] watermarks) {
		this.columns = columns;
		this.primaryKeys = primaryKeys;
		this.uniqueKeys = uniqueKeys;
		this.computedColumns = computedColumns;
		this.watermarks = watermarks;

		// validate and create name to index
		columnNameToIndex = new HashMap<>();
		final Set<String> duplicateNames = new HashSet<>();
		final Set<String> uniqueNames = new HashSet<>();
		for (int i = 0; i < this.columns.length; i++) {
			Preconditions.checkNotNull(columns[i]);
			final String fieldName = this.columns[i].name();
			columnNameToIndex.put(fieldName, i);

			if (uniqueNames.contains(fieldName)) {
				duplicateNames.add(fieldName);
			} else {
				uniqueNames.add(fieldName);
			}
		}

		if (!duplicateNames.isEmpty()) {
			throw new TableException(
				"Table column names must be unique.\n" +
				"The duplicate columns are: " + duplicateNames.toString() + "\n" +
				"All column names: " +
					Arrays.toString(Arrays.stream(this.columns).map(Column::name).toArray(String[]::new))
			);
		}

		// validate primary keys
		for (int i = 0; i < primaryKeys.length; i++) {
			if (!columnNameToIndex.containsKey(primaryKeys[i])) {
				throw new TableException(
					"Primary key field: " + primaryKeys[i] + " not found in table schema."
				);
			}
		}

		// validate unique keys
		for (int i = 0; i < uniqueKeys.length; i++) {
			String[] uniqueKey = uniqueKeys[i];

			if (null == uniqueKey || 0 == uniqueKey.length) {
				throw new TableException("Unique key should not be empty.");
			}

			for (int j = 0; j < uniqueKey.length; j++) {
				if (!columnNameToIndex.containsKey(uniqueKey[j])) {
					throw new TableException(
						"Unique key field: " + uniqueKey[j] + " not found in table schema."
					);
				}
			}
		}
	}

	public TableSchema(String[] names, InternalType[] types, boolean[] nulls) {
		this(validate(names, types, nulls));
	}

	public TableSchema(String[] names, InternalType[] types) {
		this(validate(names, types));
	}

	private static Column[] validate(String[] names, InternalType[] types, boolean[] nulls) {
		if (names.length != types.length) {
			throw new TableException(
				"Number of column indexes and column names must be equal.\n" +
					"Column names count is [" + names.length + "]\n" +
					"Column types count is [" + types.length + "]\n" +
					"Column names: " + Arrays.toString(names) + "\n" +
					"Column types: " + Arrays.toString(types)
			);
		}

		if (names.length != nulls.length) {
			throw new TableException(
				"Number of column names and nullabilities must be equal.\n" +
					"Column names count is: " + names.length + "\n" +
					"Column nullabilities count is: " + nulls.length + "\n" +
					"List of all field names: " + Arrays.toString(names) + "\n" +
					"List of all field nullabilities: " + Arrays.toString(nulls)
			);
		}

		List<Column> columns = new ArrayList<>();
		for (int i = 0; i < names.length; i++) {
			columns.add(new Column(names[i], types[i], nulls[i]));
		}
		return columns.toArray(new Column[columns.size()]);
	}

	private static Column[] validate(String[] names, InternalType[] types) {
		boolean[] nulls = new boolean[types.length];
		for (int i = 0; i < types.length; i++) {
			nulls[i] = !TimestampType.ROWTIME_INDICATOR.equals(types[i])
				&& !TimestampType.PROCTIME_INDICATOR.equals(types[i]);
		}
		return validate(names, types, nulls);
	}

	public Column[] getColumns() {
		return this.columns;
	}

	public String[] getPrimaryKeys() {
		return this.primaryKeys;
	}

	public String[][] getUniqueKeys() {
		return this.uniqueKeys;
	}

	/**
	 * Returns all field type information as an array.
	 */
	public InternalType[] getFieldTypes() {
		return Arrays.stream(columns).map(Column::internalType).toArray(InternalType[]::new);
	}

	/**
	 * Returns the specified type information for the given field index.
	 *
	 * @param fieldIndex the index of the field
	 */
	public Optional<InternalType> getFieldType(int fieldIndex) {
		if (fieldIndex < 0 || fieldIndex >= columns.length) {
			return Optional.empty();
		}
		return Optional.of(columns[fieldIndex].internalType());
	}

	/**
	 * Returns the specified type information for the given field name.
	 *
	 * @param fieldName the name of the field
	 */
	public Optional<InternalType> getFieldType(String fieldName) {
		if (columnNameToIndex.containsKey(fieldName)) {
			return Optional.of(columns[columnNameToIndex.get(fieldName)].internalType());
		}
		return Optional.empty();
	}

	/**
	 * Returns the number of fields.
	 */
	public int getFieldCount() {
		return columns.length;
	}

	/**
	 * Returns all field names as an array.
	 */
	public String[] getFieldNames() {
		return Arrays.stream(columns).map(Column::name).toArray(String[]::new);
	}

	/**
	 * Returns the specified name for the given field index.
	 *
	 * @param fieldIndex the index of the field
	 */
	public Optional<String> getFieldName(int fieldIndex) {
		if (fieldIndex < 0 || fieldIndex >= columns.length) {
			return Optional.empty();
		}
		return Optional.of(columns[fieldIndex].name());
	}

	/**
	 * Returns all field nullables as an array.
	 */
	public boolean[] getFieldNullables() {
		boolean[] nulls = new boolean[columns.length];
		for (int i = 0; i < columns.length; i++) {
			nulls[i] = columns[i].isNullable();
		}
		return nulls;
	}

	/**
	 * @deprecated Use {@link TableSchema#getFieldTypes()} instead. Can be dropped after 1.7.
	 */
	@Deprecated
	public InternalType[] getTypes() {
		return Arrays.stream(columns).map(Column::internalType).toArray(InternalType[]::new);
	}

	/**
	 * @deprecated Use {@link TableSchema#getFieldNames()} instead. Can be dropped after 1.7.
	 */
	@Deprecated
	public String[] getColumnNames() {
		return Arrays.stream(columns).map(Column::name).toArray(String[]::new);
	}

	/**
	 * @deprecated Use {@link TableSchema#getFieldNullables()} instead. Can be dropped after 1.7.
	 */
	@Deprecated
	public boolean[] getNullables() {
		return getFieldNullables();
	}

	public ComputedColumn[] getComputedColumns() {
		return computedColumns;
	}

	public Watermark[] getWatermarks() {
		return watermarks;
	}

	/**
	 * Returns the specified column for the given field index.
	 *
	 * @param fieldIndex the index of the field
	 */
	public Column getColumn(int fieldIndex) {
		Preconditions.checkArgument(fieldIndex >= 0 && fieldIndex < columns.length);
		return columns[fieldIndex];
	}

	/**
	 * Returns the map for column name to field index.
	 */
	public Map<String, Integer> columnNameToIndex() {
		return columnNameToIndex;
	}

	/**
	 * @deprecated Use {@link TableSchema#getFieldType(int)} instead. Can be dropped after 1.7.
	 */
	@Deprecated
	public InternalType getType(int fieldIndex) {
		Preconditions.checkArgument(fieldIndex >= 0 && fieldIndex < columns.length);
		return columns[fieldIndex].internalType();
	}

	/**
	 * @deprecated Use {@link TableSchema#getFieldType(String)} instead. Can be dropped after 1.7.
	 */
	@Deprecated
	public Optional<InternalType> getType(String fieldName) {
		if (columnNameToIndex.containsKey(fieldName)) {
			return Optional.of(columns[columnNameToIndex.get(fieldName)].internalType());
		}
		return Optional.empty();
	}

	/**
	 * @deprecated Use {@link TableSchema#getFieldName(int)} instead. Can be dropped after 1.7.
	 */
	@Deprecated
	public String getColumnName(int fieldIndex) {
		Preconditions.checkArgument(fieldIndex >= 0 && fieldIndex < columns.length);
		return columns[fieldIndex].name();
	}

	@Override
	public String toString() {
		final StringBuilder sb = new StringBuilder();
		sb.append("root\n");
		for (int i = 0; i < columns.length; i++) {
			sb.append(" |-- name: ").append(columns[i].name()).append("\n");
			sb.append(" |-- type: ").append(columns[i].internalType()).append("\n");
			sb.append(" |-- isNullable: ").append(columns[i].isNullable()).append("\n");
		}

		if (primaryKeys.length > 0) {
			sb.append("primary keys\n");
			sb.append(" |-- ").append(String.join(", ", primaryKeys)).append("\n");
		}

		if (uniqueKeys.length > 0) {
			sb.append("unique keys\n");
			for (int i = 0; i < uniqueKeys.length; i++) {
				sb.append(" |-- ").append(String.join(", ", uniqueKeys[i])).append("\n");
			}
		}
		return sb.toString();
	}

	@Override
	public boolean equals(Object o) {
		if (this == o) {
			return true;
		}
		if (o == null || getClass() != o.getClass()) {
			return false;
		}
		TableSchema schema = (TableSchema) o;
		return Arrays.equals(columns, schema.columns) &&
			Arrays.equals(primaryKeys, schema.primaryKeys) &&
			Arrays.equals(uniqueKeys, schema.uniqueKeys);
	}

	@Override
	public int hashCode() {
		int result = Arrays.hashCode(columns);
		result = 31 * result + Arrays.hashCode(primaryKeys);
		result = 31 * result + Arrays.hashCode(uniqueKeys);
		return result;
	}

	public static Builder builder() {
		return new Builder();
	}

	// --------------------------------------------------------------------------------------------

	/**
	 * Builder for creating a {@link TableSchema}.
	 */
	public static class Builder {

		private List<Column> columns;

		private List<String> primaryKey;

		private List<List<String>> uniqueKeys;

		private List<ComputedColumn> computedColumns;

		private List<Watermark> watermarks;

		public Builder() {
			columns = new ArrayList<>();
			primaryKey = new ArrayList<>();
			uniqueKeys = new ArrayList<>();
			computedColumns = new ArrayList<>();
			watermarks = new ArrayList<>();
		}

		/**
		 * Add a field with name and type. The call order of this method determines the order
		 * of fields in the schema.
		 */
		@Deprecated
		public Builder field(String name, InternalType type) {
			columns.add(new Column(name, type));
			return this;
		}

		@Deprecated
		public Builder field(String name, InternalType type, boolean nullable) {
			columns.add(new Column(name, type, nullable));
			return this;
		}

		public Builder column(String name, InternalType type) {
			columns.add(new Column(name, type));
			return this;
		}

		public Builder column(String name, InternalType type, boolean nullable) {
			columns.add(new Column(name, type, nullable));
			return this;
		}

		public Builder primaryKey(String... fields) {
			Arrays.stream(fields).forEach(field -> primaryKey.add(field));
			return this;
		}

		public Builder uniqueKey(String... fields) {
			List<String> uniqueKey = new ArrayList<>();
			Arrays.stream(fields).forEach(field -> uniqueKey.add(field));
			uniqueKeys.add(uniqueKey);
			return this;
		}

		public Builder computedColumn(String name, String expression) {
			computedColumns.add(new ComputedColumn(name, expression));
			return this;
		}

		public Builder watermark(String name, String eventTime, long offset) {
			watermarks.add(new Watermark(name, eventTime, offset));
			return this;
		}

		/**
		 * Returns a {@link TableSchema} instance.
		 */
		public TableSchema build() {
			return new TableSchema(
				columns.stream().toArray(Column[]::new),
				primaryKey.stream().toArray(String[]::new),
				// List<List<String>> -> String[][]
				uniqueKeys.stream()
					.map(u -> u.toArray(new String[u.size()]))  // mapping each List to an array
					.collect(Collectors.toList())               // collecting as a List<String[]>
					.toArray(new String[uniqueKeys.size()][]),
				computedColumns.stream().toArray(ComputedColumn[]::new),
				watermarks.stream().toArray(Watermark[]::new));
		}
	}
}
